Mathlib/Data/Bitvec/Core.lean

```diff
@@ -108,23 +109,23 @@ section Bitwise
 variable {n : ℕ}
 
 /-- bitwise not -/
-def not : Bitvec n → Bitvec n :=
-  map not
+def not (bv : Bitvec n) : Bitvec n :=
+  map _root_.not bv
```

Comment:
Yes. The whole point of `nonrec` is to disambiguate names though as far as I understand.

Reply:
That should be the task of namespaces

Reply:
More specifically features like this were introduced to allow shadowing of names in ways that would not trigger recursion, not as a substitute for doing the job of namespaces I.e. proper scoping and disambiguation of names.

Reply:
Based on [the zulip thread where the feature was added](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/synport.20experimentation/near/250236474), the whole point of it is to avoid namespace issues like this

Reply:
I looked further down the thread. `nonrec` was useful for disambiguating the use of `min` when there is a seemingly recursive reference of `min`. Using  `_root_.min` would not help according to the thread  (https://leanprover-community.github.io/archive/stream/287929-mathlib4/topic/synport.20experimentation.html#250239057). 

In general a feature like this (`let` vs `let_rec` in scheme and the French ML family) exists because there is a need to have the ability to shadow or not shadow the name of the definition itself. In these cases the order of evaluation can be different and the type checker can do things differently (whether or not lean4 *currently* does so is another matter) (see: .https://cs.stackexchange.com/a/101174). 

Here the issue is lean cannot tell which `not` to use, so it defaults to the local one. This is a namespace issue. Fixing it with something that is/can alter evaluation/type checking semantics is not a future proof way to solve this problem.

I don't mind making this change now. But it is a conceptually overloaded fix for a simple problem that goes away by clarifying the namespace.

Reply:
> I looked further down the thread. `nonrec` was useful for disambiguating the use of `min` when there is a seemingly recursive reference of `min`. Using `_root_.min` would not help according to the thread (https://leanprover-community.github.io/archive/stream/287929-mathlib4/topic/synport.20experimentation.html#250239057). This is not the situation here.

I'm struggling to see hwo the situation is any different there vs here. Can you quote the message that claims `_root_.min` doesn't work? (I think maybe the spelling that does work is `linear_order.min`, but that feels analogous to `_root_.min`).

Reply:
The claim is made here: https://leanprover-community.github.io/archive/stream/287929-mathlib4/topic/synport.20experimentation.html#250239310

Reply:
In this case I think actually `nonrec` is a distraction; the right thing to do is mark these definitions `protected`, since they didn't clash with anything in the `_root_` namespace in Lean 3, but now they do in Lean 4.